### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.AffineSpace.AffineMap

### PR DESCRIPTION
**Pull request by mcdoll**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mcdoll** [comment]
Blocked by leanprover/lean4#2074

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 1]
```diff
@@ -21,58 +21,72 @@ This file defines affine maps.
...
-
 open Affine
 
-/-- An `affine_map k P1 P2` (notation: `P1 →ᵃ[k] P2`) is a map from `P1` to `P2` that
+section defn
+
+variable (k : Type _) {V1 : Type _} (P1 : Type _) {V2 : Type _} (P2 : Type _) [Ring k]
+  [AddCommGroup V1] [Module k V1] [AffineSpace V1 P1] [AddCommGroup V2] [Module k V2]
+  [AffineSpace V2 P2]
+-- Workaround for lean4#2074
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:56]:
```suggestion
-- Porting note: Workaround for lean4#2074
```

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 2]
```diff
@@ -21,58 +21,72 @@ This file defines affine maps.
...
+    [AddCommGroup V1] [Module k V1] [AffineSpace V1 P1] [AddCommGroup V2] [Module k V2]
+    [AffineSpace V2 P2] : CoeFun (P1 →ᵃ[k] P2) fun _ => P1 → P2 :=
   ⟨AffineMap.toFun⟩
 
 namespace LinearMap
 
 variable {k : Type _} {V₁ : Type _} {V₂ : Type _} [Ring k] [AddCommGroup V₁] [Module k V₁]
   [AddCommGroup V₂] [Module k V₂] (f : V₁ →ₗ[k] V₂)
 
+-- Workaround for lean4#2074
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:87]:
```suggestion
-- Porting note: Workaround for lean4#2074
```

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 3]
```diff
@@ -96,23 +110,25 @@ namespace AffineMap
 
 variable {k : Type _} {V1 : Type _} {P1 : Type _} {V2 : Type _} {P2 : Type _} {V3 : Type _}
   {P3 : Type _} {V4 : Type _} {P4 : Type _} [Ring k] [AddCommGroup V1] [Module k V1]
-  [affine_space V1 P1] [AddCommGroup V2] [Module k V2] [affine_space V2 P2] [AddCommGroup V3]
-  [Module k V3] [affine_space V3 P3] [AddCommGroup V4] [Module k V4] [affine_space V4 P4]
+  [AffineSpace V1 P1] [AddCommGroup V2] [Module k V2] [AffineSpace V2 P2] [AddCommGroup V3]
+  [Module k V3] [AffineSpace V3 P3] [AddCommGroup V4] [Module k V4] [AffineSpace V4 P4]
 
-include V1 V2
+-- Workaround for lean4#2074
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:116]:
```suggestion
-- Porting note: Workaround for lean4#2074
```

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 4]
```diff
@@ -236,37 +259,49 @@ theorem smul_linear (t : R) (f : P1 →ᵃ[k] V2) : (t • f).linear = t • f.l
   rfl
 #align affine_map.smul_linear AffineMap.smul_linear
 
-instance [DistribMulAction Rᵐᵒᵖ V2] [IsCentralScalar R V2] : IsCentralScalar R (P1 →ᵃ[k] V2)
-    where op_smul_eq_smul r x := ext fun _ => op_smul_eq_smul _ _
+variable [DistribMulAction Rᵐᵒᵖ V2] [IsCentralScalar R V2]
+
+-- Workaround for lean4#2074
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:264]:
```suggestion
-- Porting note: Workaround for lean4#2074
```

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 5]
```diff
@@ -492,23 +517,24 @@ theorem linear_bijective_iff (f : P1 →ᵃ[k] P2) :
...
   · rintro ⟨-, ⟨x, hx, y, hy, rfl⟩, rfl⟩
-    exact ⟨x, hx, y, hy, rfl⟩
+    exact ⟨x, y, hx, hy, rfl⟩
 #align affine_map.image_vsub_image AffineMap.image_vsub_image
 
-omit V2
-
 /-! ### Definition of `affine_map.line_map` and lemmas about it -/
 
+-- Workaround for lean4#2074
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:532]:
```suggestion
-- Porting note: Workaround for lean4#2074
```

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 6]
```diff
@@ -758,11 +777,19 @@ variable [Semiring R] [Module R V2] [SMulCommClass k R V2]
...
   { AffineMap.distribMulAction with
     smul := (· • ·)
-    add_smul := fun c₁ c₂ f => ext fun p => add_smul _ _ _
-    zero_smul := fun f => ext fun p => zero_smul _ _ }
+    add_smul := fun _ _ _ => ext fun _ => add_smul _ _ _
+    zero_smul := fun _ => ext fun _ => zero_smul _ _ }
 
 variable (R)
 
+-- Workarounds for lean4#2074
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:785]:
```suggestion
-- Porting note: Workarounds for lean4#2074
```

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 7]
```diff
@@ -758,11 +777,19 @@ variable [Semiring R] [Module R V2] [SMulCommClass k R V2]
...
+instance : AddCommMonoid (V1 →ₗ[k] V2) := LinearMap.addCommMonoid
+instance : AddCommMonoid (V2 × (V1 →ₗ[k] V2)) := Prod.instAddCommMonoidSum
+instance : Module R (V1 →ₗ[k] V2) := LinearMap.instModuleLinearMapAddCommMonoid
+instance : Module R (V2 × (V1 →ₗ[k] V2)) := Prod.module
+-- Workaround for lean4#2074
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:790]:
```suggestion
-- Porting note: Workaround for lean4#2074
```

File: Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean [hunk 8]
```diff
@@ -0,0 +1,919 @@
...
+
+/-- An `AffineMap k P1 P2` (notation: `P1 →ᵃ[k] P2`) is a map from `P1` to `P2` that
+induces a corresponding linear map from `V1` to `V2`. -/
+structure AffineMap
+  where
+  /-- The affine map as a function-/
+  toFun : P1 → P2
+  /-- The underlying linear map -/
+  linear : V1 →ₗ[k] V2
+  /-- Affineness -/
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/AffineSpace/AffineMap.lean:67]:
```suggestion
structure AffineMap where
  toFun : P1 → P2
  linear : V1 →ₗ[k] V2
```
The docs linter can (and arguably should) be ignored while porting.

**Comment by jcommelin** [comment]
bors d+

**Comment by mcdoll** [comment]
bors merge

**Pull request closed without merging**